[TAN-5175] AI-generated file descriptions

adessy opened the pull request:
# Changelog
## Added
- [TAN-5175] Implements AI-generated file descriptions functionality, adding the capability to automatically generate multilingual descriptions for uploaded files using OpenAI's GPT-4.1 model.

adessy committed:
[TAN-5175] Implement a first, basic version of the description generator

adessy committed:
[TAN-5175] Add tracking to DescriptionGenerationJob

adessy committed:
[TAN-5175] Prevent the file description generator to overwrite existing descriptions

adessy committed:
[TAN-5175] Use structured outputs when generating file descriptions

Use JSON Schema to specify the structure of the LLM response, instead of relying on prompt engineering.

adessy committed:
[TAN-5175] Remove unused `Files::DescriptionGenerator#validate_file!`

We might reintroduce it later if there is a need for further validation.

adessy committed:
[TAN-5175] Add tests for Files::DescriptionGenerator

adessy committed:
[TAN-5175] Remove outdated TODO comment in Files::File model

adessy committed:
[TAN-5175] Add tests for Files::File#description_generation_status

adessy committed:
[TAN-5175] Add tests for Files::DescriptionGenerationJob

adessy committed:
[TAN-5175] Add preprocessing and support for additional file types in description generation

Includes handling of previewable PDF files and large images.

* Large files are resized unless they are too big, in which case description generation is interrupted.
* Uses the PDF preview, if available, for files not natively supported by the LLM.
* Extra fallback measure for DOCX files: if preview creation fails for a DOCX file, it's converted to HTML on the fly and sent to the LLM.

adessy committed:
[TAN-5175] Include file name in description generation prompt

adessy committed:
[TAN-5175] Minor improvements to file description generation

kogre commented on back/Gemfile.lock:
Thanks for the upgrades :)

kogre commented on back/app/jobs/files/description_generation_job.rb:
```diff
@@ -0,0 +1,48 @@
+    end
+
+    def job_tracking_context
+      arguments.first.tap do |file|
+        raise "Expected file, got #{file.class}" unless file.is_a?(Files::File)
+      end
```
Not important and nothing wrong with this, it's just cognitive lower effort to read (imo)  `file = arguments.first`

kogre commented on back/app/services/files/description_generator.rb:
```diff
@@ -0,0 +1,199 @@
+    end
+
+    def previewable_as_pdf?(file)
+      PreviewService.new.send(:should_generate_preview?, file)
```
Let's make `should_generate_preview?` public?

kogre commented on back/app/services/files/description_generator.rb:
```diff
@@ -0,0 +1,199 @@
+      if preview.content.full_url.present?
+        preview.content
+      elsif preview.status == 'pending'
+        raise ApplicationJob::RetryInError.new('Preview is pending, try again later.', 30.seconds)
```
I think this is okay, but 2 thoughts I want to share:
- Not 100% sure we want the retry logic to leak into the service. I guess it's fine because the interface is minimal and it's pretty convenient.
- If wanting to be true to the Ruby conventions (and some technical differences I honestly forgot), it's better to use `throw/catch` instead of `raise/rescue` for deliberate control flow, as opposed to error handling.

adessy replied:
I agree with point (1), but not with point (2). In my opinion, `throw/catch` is a bit of a misfeature and, if used, should better remain confined within the method. I'll replace it with another exception.

kogre commented on back/app/services/files/description_generator.rb:
```diff
@@ -0,0 +1,199 @@
+    end
+
+    # @return [Tempfile] Tempfile with HTML content
+    def docx_to_html(file)
+      Tempfile.open do |local_file|
+        local_file.binmode.write(file.content.read)
+        doc = Docx::Document.open(local_file.path)
+
+        Tempfile
+          .new([file.name, '.html'])
+          .tap { |processed_file| processed_file.write(doc.to_html) }
+      end
+    end
```
Not super sure whether we really need this, and the extra dependency that comes with it. Are you sure this is a meaningful improvement over the preview?

adessy replied:
Me neither, this was actually implemented before using the preview. I'll leave it for now to save some time.

kogre commented on back/app/services/files/description_generator.rb:
```diff
@@ -0,0 +1,199 @@
+        The description should be informative and capture the essential meaning of the document.
+
+        Requirements:
+        - Avoid starting the summary with "The document is..." or similar generic introductions if possible
```
Better to use `description` instead of `summary` to be consistent with the earlier terminology and not to confuse the model?

adessy replied:
Good catch

kogre commented on back/app/services/files/description_generator.rb:
```diff
@@ -0,0 +1,199 @@
+      #
+      # @param file [Files::File] The file to generate descriptions for
+      # @return [Boolean] true if a job was enqueued, false otherwise.
+      def enqueue_job(file)
```
Just curious, why not make these methods instance methods as well?

adessy replied:
No good or bad reason. It just felt more like a factory method.

kogre commented on back/config/initializers/ruby_llm.rb:
```diff
@@ -0,0 +1,5 @@
+
+RubyLLM.configure do |config|
+  config.openai_api_key = ENV.fetch('OPENAI_API_KEY', nil)
+end
```
Making sure we use Azure

kogre reviewed (approved):
Looks fine, some questions/challenges, but nothing locking 👏 

I've faced some similar challenges in the transcription PR (waiting on processing and factories running jobs come to mind), might be good to align both tracks and use the same fundamentals and solutions, and yours look more polished. But not sure I'll have the time.

adessy committed:
[TAN-5175] Make `Files::PreviewService#should_generate_preview?` method public

adessy committed:
[TAN-5175] Remove `RetryInError` and introduce a specific `PreviewPendingError` in the description generator

The `RetryInError` error encourages retry logic to leak into other parts of the codebase, like services, where it doesn’t belong.

adessy committed:
[TAN-5175] Consistently use "description" in `DescriptionGenerator` prompt

adessy committed:
[TAN-5175] Rework RubyLLM config to use Azure OpenAI

adessy committed:
[TAN-5175] Bump RubyLLM version 1.4.2 to 1.6.2

adessy committed:
[TAN-5175] Replace GPT 4.1 with Claude models for file description generation

We had to pause our use of GPT for the generation of file descriptions because the first implementation relied on GPT from OpenAI. However, OpenAI is not one of the data providers permitted under our DPA. We should instead use GPT models as served by Azure. Unfortunately, when it comes to file support, Azure OpenAI shows some parity issues compared to OpenAI APIs.

After a lot of tinkering and attempts to make it work with the new Responses API (instead of the chat completion API), we managed to get it running to some extent, but only with significant extra rework, and the results still fell short in terms of quality.

As a result, we decided to switch to using Claude on Bedrock instead. While it comes with its own limitations, it requires less refactoring of the original solution and seems to produce better results.

Shortcomings:
- Claude models do not support structured output, so we had to rely on prompt engineering and response prefilling to guide the model toward producing proper JSON replies.
- PDF support is handled in text mode, which means images in PDFs are ignored. To get full visual understanding, we would need to enable citations, but this is not directly supported in RubyLLM and would change the structure of the output.

adessy committed:
Merge branch 'master' into TAN-5175/file-description-autogenerated

# Conflicts:
#	back/db/structure.sql

kogre commented on back/app/services/files/description_generator.rb:
```diff
@@ -40,13 +40,43 @@ def tenant_locales
-      prompt = build_prompt(file.name)
+      chat = RubyLLM.chat(model: infer_best_model)
+      prompt = build_prompt(file.name, locales)
+      prefill_msg = '{' # Prefill the response to encourage the LLM to respond with a JSON object
```
Tiny little detail, but you can already do `{"` to increase chances even more

kogre reviewed (approved):
Looks good to me :)
